### PR TITLE
feat: add fs-repo-migration-12-to-13

### PR DESCRIPTION
**Pull request by Jorropo**
Unblock: https://github.com/ipfs/kubo/pull/9492

### CODE REVIEW

**Review by galargh — COMMENTED** [review]
In comparison to https://github.com/ipfs/distributions/tree/master/dists/fs-repo-11-to-12 we're missing:
- `no-site`
- `repo-owner`
- `vtag`
files. I'm not 100% whether they're optional or required. I think it'd be quicker to add them. Would you mind doing that?

**Comment by Jorropo** [comment]
@galargh I'll do, just so you know I just did a `./dist.sh new-go-dist fs-repo-12-to-13 github.com/ipfs/fs-repo-migrations` and then fixed the versions and `package =` lines.

**Review by galargh — APPROVED** [review]

**Comment by Jorropo** [comment]
Notes for self:
All amd64 microarchitectures significantly increase build times.
For future fs-repo-* we probably don't need them (as usually less than <1% performance improvements aren't key in the migration). For daemons like kubo it's usefull tho.

Also concurrent builds would be nice.

File: templates/build_matrix [hunk 1]
```diff
@@ -1,4 +1,6 @@
 darwin amd64
+darwin amd64-v2
+darwin amd64-v3
```
[the remark this answers is not in the document]
Comment (Jorropo) [reply templates/build_matrix:3]: right, I'll remove them.

Since I'll have to run CI again anyway, I'll remove all v2 and v3 for fs-repo-* (see https://github.com/ipfs/distributions/pull/790#issuecomment-1346863565), the goal is to improve daemons like Kubo.

File: templates/build_matrix [hunk 2]
```diff
@@ -18,3 +18,5 @@ linux arm
 linux arm64
 windows 386
 windows amd64
+windows amd64-v2
+windows amd64-v3
```
Comment (lidel) [inline templates/build_matrix:22]: @Jorropo mind moving this unrelated change to separate PR? 

I feel wee need to discuss if there is any point in additional build time for windows, if IPFS DEsktop always uses the `amd64` one.

Comment (Jorropo) [reply templates/build_matrix:22]: I was expecting to update ipfs-update and maybe ipfs-desktop later.
Anyway will done: https://github.com/ipfs/distributions/pull/791

**Review by lidel — CHANGES_REQUESTED** [review]
Something went wrong, invalid binary is in the linux package:

```console
$ tar xf fs-repo-12-to-13_v1.0.0_linux-amd64.tar.gz
fs-repo-12-to-13/fs-repo-11-to-12
$ chmod +x fs-repo-12-to-13/fs-repo-11-to-12                                                                                        
$  fs-repo-12-to-13/fs-repo-11-to-12  --help
panic: qtls.ClientHelloInfo doesn't match

goroutine 1 [running]:
github.com/marten-seemann/qtls-go1-15.init.0()
	github.com/marten-seemann/qtls-go1-15@v0.1.1/unsafe.go:20 +0x132
```

File: dists/fs-repo-12-to-13/Makefile [hunk 3]
```diff
@@ -0,0 +1,5 @@
+repo = github.com/ipfs/fs-repo-migrations
+# sub-package to build (omit when building the main package)
+package = fs-repo-11-to-12
```
Comment (lidel) [inline dists/fs-repo-12-to-13/Makefile:3]:
```suggestion
package = fs-repo-12-to-13
```

**Comment by Jorropo** [comment]
Works for me:
```
Initializing daemon...
Kubo version: 0.18.0-dev
Repo version: 13
System version: amd64/linux
Golang version: go1.19.4
Found outdated fs-repo, migrations need to be run.
Looking for suitable migration binaries.
Need 1 migrations, downloading.
Downloading migration: fs-repo-12-to-13...
Fetching with HTTP: "https://ipfs.io/ipfs/QmZwD9CY1YxKDvecfkZgPP5DAr3nbvvH3cxpD8rioNjiwW/fs-repo-12-to-13/versions"
Fetching with HTTP: "https://ipfs.io/ipfs/QmZwD9CY1YxKDvecfkZgPP5DAr3nbvvH3cxpD8rioNjiwW/fs-repo-12-to-13/versions"
Fetching with HTTP: "https://ipfs.io/ipfs/QmZwD9CY1YxKDvecfkZgPP5DAr3nbvvH3cxpD8rioNjiwW/fs-repo-12-to-13/v1.0.0/fs-repo-12-to-13_v1.0.0_linux-amd64.tar.gz"
Downloaded and unpacked migration: /tmp/migrations2492603280/fs-repo-12-to-13 (v1.0.0)
Running migration fs-repo-12-to-13 ...
  => Running: /tmp/migrations2492603280/fs-repo-12-to-13 -path=/home/hugo/.ipfs -verbose=true
applying 12-to-13 repo migration
locking repo at "/home/hugo/.ipfs"
  - verifying version is '12'
> Upgrading config to new format
Routing.Type settings is different than the old default, skipping
Reprovider settings are different than the old default, skipping
Swarm.ConnMgr settings are different than the old defaults, skipping
updated version file
Migration 12 to 13 succeeded
Success: fs-repo migrated to version 13.

Computing default go-libp2p Resource Manager limits based on:
    - 'Swarm.ResourceMgr.MaxMemory': "13 GB"
    - 'Swarm.ResourceMgr.MaxFileDescriptors': 524288

Applying any user-supplied overrides on top.
Run 'ipfs swarm limit all' to see the resulting limits.

Swarm listening on /ip4/127.0.0.1/tcp/4001
Swarm listening on /ip4/127.0.0.1/udp/4001/quic
Swarm listening on /ip4/127.0.0.1/udp/4001/quic-v1
Swarm listening on /ip4/127.0.0.1/udp/4001/quic-v1/webtransport/certhash/uEiBVdsYoVs0r-yMV-7p4vngPfpAZ49QT6gRNqeRZG_OQfw/certhash/uEiAc2l0momJCtXw6XSEi4O3RTzpX3PZc4Nvg_VKRzBs9EQ
Swarm listening on /ip4/172.17.0.1/tcp/4001
Swarm listening on /ip4/172.17.0.1/udp/4001/quic
Swarm listening on /ip4/172.17.0.1/udp/4001/quic-v1
Swarm listening on /ip4/172.17.0.1/udp/4001/quic-v1/webtransport/certhash/uEiBVdsYoVs0r-yMV-7p4vngPfpAZ49QT6gRNqeRZG_OQfw/certhash/uEiAc2l0momJCtXw6XSEi4O3RTzpX3PZc4Nvg_VKRzBs9EQ
Swarm listening on /ip4/192.168.1.33/tcp/4001
Swarm listening on /ip4/192.168.1.33/udp/4001/quic
Swarm listening on /ip4/192.168.1.33/udp/4001/quic-v1
Swarm listening on /ip4/192.168.1.33/udp/4001/quic-v1/webtransport/certhash/uEiBVdsYoVs0r-yMV-7p4vngPfpAZ49QT6gRNqeRZG_OQfw/certhash/uEiAc2l0momJCtXw6XSEi4O3RTzpX3PZc4Nvg_VKRzBs9EQ
Swarm listening on /ip6/204:e3e5:ec57:9482:3f92:bf4d:74bf:cfed/tcp/4001
Swarm listening on /ip6/204:e3e5:ec57:9482:3f92:bf4d:74bf:cfed/udp/4001/quic
Swarm listening on /ip6/204:e3e5:ec57:9482:3f92:bf4d:74bf:cfed/udp/4001/quic-v1
Swarm listening on /ip6/204:e3e5:ec57:9482:3f92:bf4d:74bf:cfed/udp/4001/quic-v1/webtransport/certhash/uEiBVdsYoVs0r-yMV-7p4vngPfpAZ49QT6gRNqeRZG_OQfw/certhash/uEiAc2l0momJCtXw6XSEi4O3RTzpX3PZc4Nvg_VKRzBs9EQ
Swarm listening on /ip6/2a01:cb05:806e:900:bf52:4b68:3b97:aa23/tcp/4001
Swarm listening on /ip6/2a01:cb05:806e:900:bf52:4b68:3b97:aa23/udp/4001/quic
Swarm listening on /ip6/2a01:cb05:806e:900:bf52:4b68:3b97:aa23/udp/4001/quic-v1
Swarm listening on /ip6/2a01:cb05:806e:900:bf52:4b68:3b97:aa23/udp/4001/quic-v1/webtransport/certhash/uEiBVdsYoVs0r-yMV-7p4vngPfpAZ49QT6gRNqeRZG_OQfw/certhash/uEiAc2l0momJCtXw6XSEi4O3RTzpX3PZc4Nvg_VKRzBs9EQ
Swarm listening on /ip6/2a01:cb05:806e:900:e965:d65b:eb5f:6ac1/tcp/4001
Swarm listening on /ip6/2a01:cb05:806e:900:e965:d65b:eb5f:6ac1/udp/4001/quic
Swarm listening on /ip6/2a01:cb05:806e:900:e965:d65b:eb5f:6ac1/udp/4001/quic-v1
Swarm listening on /ip6/2a01:cb05:806e:900:e965:d65b:eb5f:6ac1/udp/4001/quic-v1/webtransport/certhash/uEiBVdsYoVs0r-yMV-7p4vngPfpAZ49QT6gRNqeRZG_OQfw/certhash/uEiAc2l0momJCtXw6XSEi4O3RTzpX3PZc4Nvg_VKRzBs9EQ
Swarm listening on /ip6/::1/tcp/4001
Swarm listening on /ip6/::1/udp/4001/quic
Swarm listening on /ip6/::1/udp/4001/quic-v1
Swarm listening on /ip6/::1/udp/4001/quic-v1/webtransport/certhash/uEiBVdsYoVs0r-yMV-7p4vngPfpAZ49QT6gRNqeRZG_OQfw/certhash/uEiAc2l0momJCtXw6XSEi4O3RTzpX3PZc4Nvg_VKRzBs9EQ
Swarm listening on /p2p-circuit
Swarm announcing /ip4/1.2.3.4/udp/1337/quic
Swarm announcing /ip4/1.2.3.4/udp/1337/quic-v1
Swarm announcing /ip4/1.2.3.4/udp/1337/quic-v1/webtransport
API server listening on /ip4/127.0.0.1/tcp/5001
WebUI: http://127.0.0.1:5001/webui
Added migration file "fs-repo-12-to-13_v1.0.0_linux-amd64.tar.gz": /ipfs/QmbiCWGuBHrheq5TkairJCwXrka9xpWaFv5CMy64zucB2w
Could not add migration to IPFS: nothing downloaded by ipfs fetcher
Gateway (readonly) server listening on /ip4/127.0.0.1/tcp/8080
Daemon is ready
```

**Review by lidel — APPROVED** [review]
Thanks, lgtm. 

Merging. After it builds in `master` and updates DNSLink, you can can bump CID in https://github.com/ipfs/kubo/pull/9492
